### PR TITLE
KAFKA-7052 Avoiding NPE in ExtractField SMT in case of non-existent fields

### PR DESCRIPTION
**Pull request by gunnarmorling**
… in case of non-existent fields

https://issues.apache.org/jira/browse/KAFKA-7052

*More detailed description of your change
n/a

*Summary of testing strategy (including rationale)
Added JUnit test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

@rhauch, hey, that's a quick first attempt for making the `ExtractField` SMT more flexible when it comes to encountering a record that doesn't contain the specified field. It's a common situation for connectors like Debezium which produce different "kinds" of records/topics; in our case e.g. actual change event topics and meta-topics such as TX data or schema history. One might want to apply the `ExtractField` SMT to the CDC records but not to those others; as that's currently not possible, a strategy is needed for more gracefully handling records without the specified field. Also see [KAFKA-7052](https://issues.apache.org/jira/browse/KAFKA-7052) for some backgrounds.

The proposal is to add a new option `behavior.on.non.existent.field` to the SMT which makes the behavior configurable. Its supported values are:

* fail: raise an exception (default for records with schema)
* return-null: return null (default for records without schema)
* pass-on: pass on the unmodified original record

The defaults are so to keep backwards compatibility with the current behavior. The "pass-on" value will address the original use case reported in KAFKA-7052.

I did a quick implementation of that proposal to foster feedback. Happy to adjust and expand as needed, e.g. to adjust with existing naming patterns for the option and/or its values as well as docs (not sure where that'd go). Thanks!

CC @rmoff, @big-andy-coates.

### CODE REVIEW

**Review by rhauch — CHANGES_REQUESTED** [review]
Thanks for the proposed change, @gunnarmorling.

As it stands, this change will require a KIP because it is altering the public API (the configuration properties) for the ExtractField SMT. As noted on [KAFKA-7052](https://issues.apache.org/jira/browse/KAFKA-7052), I think we can actually improve the error message and possibly change the implementation (without introducing a new config property) to skip any record for which there is no specified field -- this wouldn't require a KIP because the NPE would be an unintentional UX and therefore a bug.

So, as-is we cannot merge this PR without an accompanying KIP that has been approved, so I've marked this PR as "Request changes" just to prevent a committer from inadvertently merging this PR before we have such a KIP.

WDYT?

**Comment by gunnarmorling** [comment]
@rhauch, ok, I've updated this PR to only avoid the NPE and raise an IAE with a more meaningful message. I'll look into the KIP for the new option separately as you suggested.

**Review by rhauch — APPROVED** [review]
LGTM. Thanks, @gunnarmorling

**Comment by rhauch** [comment]
ok to test

**Comment by rhauch** [comment]
retest this please

**Comment by rhauch** [comment]
Test results on JDK 8 were only these unrelated and flaky tests:
```
kafka.admin.DescribeConsumerGroupTest.testDescribeGroupMembersWithShortInitializationTimeout
kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
```

**Pull request merged**
